proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,577 @@
+       TupleArgKind t_or_c 4 = Tuple4 t_or_c t_or_c t_or_c t_or_c
+       -- ...
+       TupleArgKind t_or_c 64 = Tuple64 t_or_c ... t_or_c
+       TupleArgKind _      n  = TypeError (ShowType n :<>: Text " is too large; the maximum size for a tuple is 64.")
+
+     type Tuple :: forall (n :: Nat). TupleArgKind Type n -> Type
+     type family Tuple ts where
+       Tuple () = Unit
+       Tuple a = a    -- see below
+       Tuple (a, b) = Tuple2 a b
+       Tuple (a, b, c) = Tuple3 a b c
+       Tuple (a, b, c, d) = Tuple4 a b c d
+       -- ...
+       Tuple (a, b, ..., bk, bl) = Tuple64 a b ... bk bl
+       Tuple @n _ = TypeError (ShowType n :<>: Text " is too large; the maximum size for a tuple is 64.")
+
+     type Constraints :: forall (n :: Nat). TupleArgKind Constraint n -> Constraint
+     type family Constraints ts where
+       Constraints () = Unit
+       Constraints a = a    -- see below
+       Constraints (a, b) = CTuple2 a b
+       Constraints (a, b, c) = CTuple3 a b c
+       Constraints (a, b, c, d) = CTuple4 a b c d
+       -- ...
+       Constraints (a, b, ..., bk, bl) = CTuple64 a b ... bk bl
+       Constraints @n _ = TypeError (ShowType n :<>: Text " is too large; the maximum size for a tuple is 64.")
+
+     type TupleArgKind# :: List RuntimeRep -> Type
+     type family TupleArgKind# reps where
+       TupleArgKind# [] = Unit
+       TupleArgKind# [r1] = TYPE r1     -- *not* Solo (TYPE r1); see below
+       TupleArgKind# [r1, r2] = Tuple2 (TYPE r1) (TYPE r2)
+       TupleArgKind# [r1, r2, r3] = Tuple3 (TYPE r1) (TYPE r2) (TYPE r3)
+       -- ...
+       TupleArgKind# [r1, ..., r64] = Tuple64 (TYPE r1) ... (TYPE r64)
+       TupleArgKind# other = TypeError (ShowType (Length other) :<>: Text " is too large; the maximum size of a tuple is 64.")
+
+     type Tuple# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (TupleRep reps)
+     type family Tuple# ts where
+       Tuple# () = Unit#
+       Tuple# (a :: TYPE r) = TypeError (Text "Tuple# does not work for 1-tuples; use Solo#.")  -- see below
+       Tuple# (a, b) = Tuple2# a b
+       Tuple# (a, b, c) = Tuple3# a b c
+       -- ...
+       Tuple# (a, b, ..., bk, bl) = Tuple64# a b ... bk bl
+       Tuple# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a tuple is 64.")
+
+     type Sum# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (SumRep reps)
+     type family Sum# ts where
+       Sum# () = TypeError (Text "GHC does not support empty unboxed sums. Consider Data.Void.Void instead.")
+       Sum# (a :: TYPE r) = TypeError (Text "GHC does not support unary unboxed sums. Consider Data.Tuple.Solo# instead.")
+       Sum# (a, b) = Sum2# a b
+       Sum# (a, b, c) = Sum3# a b c
+       -- ...
+       Sum# (a, b, ..., bk, bl) = Sum64# a b ... bk bl
+       Sum# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a sum is 64.")
```

Comment:
Today, unboxed sums are not limited to 64 alternatives.

```
ghci> :t (# () | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | #)
(# () | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | #)
  :: (# () | b | c | d | e | f | g | h | i | j | k | l | m | n | o |
        p | q | r | s | t | u | v | w | x | y | z | t26 | t27 | t28 | t29 |
        t30 | t31 | t32 | t33 | t34 | t35 | t36 | t37 | t38 | t39 | t40 |
        t41 | t42 | t43 | t44 | t45 | t46 | t47 | t48 | t49 | t50 | t51 |
        t52 | t53 | t54 | t55 | t56 | t57 | t58 | t59 | t60 | t61 | t62 |
        t63 | t64 | t65 | t66 | t67 | t68 | t69 | t70 | t71 | t72 | t73 |
        t74 | t75 | t76 | t77 | t78 | t79 | t80 | t81 | t82 | t83 | t84 |
        t85 | t86 | t87 | t88 | t89 | t90 | t91 | t92 | t93 | t94 | t95 |
        t96 | t97 | t98 | t99 | t100 | t101 | t102 | t103 | t104 | t105 |
        t106 | t107 | t108 | t109 | t110 | t111 | t112 | t113 | t114 | t115
        | t116 | t117 | t118 | t119 | t120 | t121 | t122 | t123 | t124 |
        t125 | t126 | t127 | t128 #)
```

That's not a major problem for this `Sum#` type family because it can't be given as input a tuple wider than 64 entries anyway. But it poses other problems:

* The `Sum<n>#` family of type constructors exported from `Data.Sum.Experimental` is not complete.
* Pretty-printing of an unboxed sum type with more than 64 alternatives cannot use the `Sum#` type family with `-XNoListTuplePuns`.
* Because the `(# | | | #)` syntax is disallowed, it is not obvious how to write a "large" unboxed sum type or term when `-XNoListTuplePuns` is active.